File: apps/desktop/src/services/client/resources/provider.ts
import type { ProviderDetail, ProviderSummary } from "../../../types";
import { call } from "../invoke";
import type { GenerateTextBody, ProviderWriteBody } from "../types";

export class ProviderResource {
  all(ws: string) {
    return call<ProviderSummary[]>("provider_list", { workspaceId: ws });
  }

  get(ws: string, id: string) {
    return call<ProviderDetail>("provider_get", { workspaceId: ws, providerId: id });
  }

  create(ws: string, body: ProviderWriteBody) {
    return call<{ id: string }>("provider_create", {
      workspaceId: ws,
      name: body.name,
      providerType: body.type,
      configDir: body.config_dir,
      description: body.description,
      mcpServers: body.mcp_servers ?? [],
    });
  }

  update(ws: string, id: string, body: ProviderWriteBody) {
    return call<void>("provider_update", {
      workspaceId: ws,
      providerId: id,
      name: body.name,
      providerType: body.type,
      configDir: body.config_dir,
      description: body.description ?? undefined,
      clearDescription: body.description === null,
      mcpServers: body.mcp_servers,
    });
  }

  generate(ws: string, body: GenerateTextBody) {
    return call<string>("provider_generate", {
      workspaceId: ws,
      providerId: body.provider_id,
      model: body.model,
      instruction: body.instruction,
      currentText: body.current_text,
      fieldLabel: body.field_label,
    });
  }

  delete(ws: string, id: string) {
    return call<void>("provider_delete", { workspaceId: ws, providerId: id });
  }
}

File: apps/desktop/src-tauri/src/grpc/provider.rs
use modula_client::{CreateProvider, CreatedProvider, GenerateText, ModulaClient, UpdateProvider};
use modula_types::{CatalogProvider, McpServer, Provider};
use tauri::State;

#[tauri::command]
pub async fn provider_catalog(
    engine: State<'_, ModulaClient>,
) -> Result<Vec<CatalogProvider>, String> {
    Ok(engine.provider_catalog().await?)
}

#[tauri::command]
pub async fn provider_list(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
) -> Result<Vec<Provider>, String> {
    Ok(engine.list_providers(&workspace_id).await?)
}

#[tauri::command]
pub async fn provider_get(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    provider_id: String,
) -> Result<Provider, String> {
    Ok(engine.get_provider(&workspace_id, &provider_id).await?)
}

#[tauri::command]
pub async fn provider_create(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    name: String,
    provider_type: String,
    config_dir: String,
    description: Option<String>,
    mcp_servers: Vec<McpServer>,
) -> Result<CreatedProvider, String> {
    Ok(engine
        .create_provider(CreateProvider {
            workspace_id,
            name,
            r#type: provider_type,
            config_dir,
            description,
            mcp_servers,
        })
        .await?)
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn provider_update(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    provider_id: String,
    name: Option<String>,
    provider_type: Option<String>,
    config_dir: Option<String>,
    description: Option<String>,
    clear_description: bool,
    mcp_servers: Option<Vec<McpServer>>,
) -> Result<(), String> {
    engine
        .update_provider(UpdateProvider {
            workspace_id,
            provider_id,
            name,
            r#type: provider_type,
            config_dir,
            description,
            clear_description,
            mcp_servers,
        })
        .await?;
    Ok(())
}

#[tauri::command]
pub async fn provider_generate(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    provider_id: String,
    model: Option<String>,
    instruction: String,
    current_text: String,
    field_label: Option<String>,
) -> Result<String, String> {
    Ok(engine
        .generate_text(GenerateText {
            workspace_id,
            provider_id,
            model,
            instruction,
            current_text,
            field_label,
        })
        .await?)
}

#[tauri::command]
pub async fn provider_delete(
    engine: State<'_, ModulaClient>,
    workspace_id: String,
    provider_id: String,
) -> Result<(), String> {
    engine.delete_provider(&workspace_id, &provider_id).await?;
    Ok(())
}
